### PR TITLE
facts - properly report virtualization facts for Linux guests on bhyve

### LINKED ISSUE #73167: Virtualization detection fails for Bhyve guests
##### SUMMARY
When i collect facts from FreeBSD bhyve target, with the command : ansible xxx -m setup | grep "ansible_virtualization_type"
the variable "ansible_virtualization_type" and "ansible_virtualization_role" is "NA".
Result of dmidecode -s system-product-name
```
dmidecode -s system-product-name
BHYVE
```

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
module_utils/facts/virtual/linux.py

##### ANSIBLE VERSION
```
ansible 2.10.4
  config file = /srv/ansible/ansible.cfg
  configured module search path = ['/srv/ansible/library']
  ansible python module location = /srv/ansible-venv/lib/python3.7/site-packages/ansible
  executable location = /srv/ansible-venv/bin/ansible
  python version = 3.7.3 (default, Dec 20 2019, 18:57:59) [GCC 8.3.0]

```

##### CONFIGURATION
```
ANSIBLE_FORCE_COLOR(/srv/ansible/ansible.cfg) = True
DEFAULT_HOST_LIST(/srv/ansible/ansible.cfg) = ['/srv/ansible/site.host']
DEFAULT_JINJA2_EXTENSIONS(/srv/ansible/ansible.cfg) = jinja2.ext.do
DEFAULT_MANAGED_STR(/srv/ansible/ansible.cfg) = Gestion par Ansible: {file} modifie le %Y-%m-%d %H:%M:%S par {uid} depuis {host}
DEFAULT_MODULE_PATH(/srv/ansible/ansible.cfg) = ['/srv/ansible/library']
DEFAULT_ROLES_PATH(/srv/ansible/ansible.cfg) = ['/srv/ansible/roles']
DEFAULT_STDOUT_CALLBACK(/srv/ansible/ansible.cfg) = yaml
DISPLAY_SKIPPED_HOSTS(/srv/ansible/ansible.cfg) = False
INTERPRETER_PYTHON(/srv/ansible/ansible.cfg) = auto_silent
RETRY_FILES_ENABLED(/srv/ansible/ansible.cfg) = False
```

##### OS / ENVIRONMENT
OS Target : Debian 10 in BHYVE hardware hosted on TrueNAS.

##### STEPS TO REPRODUCE
Create a VM in bhyve host, run the commande `ansible -m setup`

##### EXPECTED RESULTS

        "ansible_virtualization_role": "guest",
        "ansible_virtualization_type": "bhyve",

##### ACTUAL RESULTS

        "ansible_virtualization_role": "NA",
        "ansible_virtualization_type": "NA",

### PR DESCRIPTION
**Pull request by samdoran**
##### SUMMARY

Fixes #73167

Add unit tests for this specific change. We did not have any tests for this file prior, so more could be added.

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

`lib/ansible/module_utils/facts/virtual/linux.py`

### CODE REVIEW

File: lib/ansible/module_utils/facts/virtual/linux.py [hunk 1]
```diff
@@ -352,6 +352,11 @@ def get_virtual_facts(self):
                         virtual_facts['virtualization_role'] = 'guest'
                         found_virt = True
 
+                if 'BHYVE' in out:
+                    virtual_facts['virtualization_type'] = 'bhyve'
+                    virtual_facts['virtualization_role'] = 'guest'
+                    found_virt = True
```
Comment (relrod) [inline lib/ansible/module_utils/facts/virtual/linux.py]: I think we'll need something like:

```suggestion
                if 'BHYVE' in out:
                    guest_tech.add('bhyve')
                    if not found_virt:
                        virtual_facts['virtualization_type'] = 'bhyve'
                        virtual_facts['virtualization_role'] = 'guest'
                        found_virt = True
```

Basically: If there's nested or multiple virt techs at play and we found something earlier, go with that, but add bhyve to the list of guest techs no matter what.

Comment (samdoran) [reply lib/ansible/module_utils/facts/virtual/linux.py]: Will do.

**Review by relrod — APPROVED** [review]
Approved, pending feedback from original reporter.

**Comment by samdoran** [comment]
The issue author [confirmed](https://github.com/ansible/ansible/issues/73167#issuecomment-760242239) this works as expected.

**Pull request merged**